prow/plugins/branch-updater/branch-updater.go

```diff
@@ -0,0 +1,164 @@
+	GetPullRequest(owner, repo string, number int) (*github.PullRequest, error)
+	GetCombinedStatus(org, repo, ref string) (*github.CombinedStatus, error)
+}
+
+func helpProvider(config *plugins.Configuration, enabledRepos []config.OrgRepo) (*pluginhelp.PluginHelp, error) {
+	configInfo := map[string]string{}
+	for _, repo := range enabledRepos {
+		repoUpdateTmpl := "PR branches in repo %s will not be automatically updated."
+		if config.BranchUpdater.ShouldUpdateBranchesForRepo(repo.String()) {
+			repoUpdateTmpl = "PR branches in repo %s will be automatically updated."
+		}
+		configInfo[repo.String()] = fmt.Sprintf(repoUpdateTmpl, repo)
+	}
+
+	yamlSnippet, _ := plugins.CommentMap.GenYaml(&plugins.Configuration{
+		BranchUpdater: plugins.BranchUpdater{
+			DefaultEnabled: true,
+			IgnoredRepos:   []string{"db-improbable/do-not-update-me"},
+			IncludeRepos:   []string{"db-improbable/please-update-me"},
+		},
+	})
+
+	pluginHelp := &pluginhelp.PluginHelp{
+		Description: "The branch-updater plugin updates the branch when an otherwise-mergeable PR is blocked from merging because it's out of date from the HEAD branch",
+		Config:      configInfo,
+		Snippet:     yamlSnippet,
+	}
+	return pluginHelp, nil
+}
+
+func init() {
+	plugins.RegisterPullRequestHandler(pluginName, pullRequestHandler, helpProvider)
+}
+
+func pullRequestHandler(pc plugins.Agent, event github.PullRequestEvent) error {
+	return handlePR(pc.GitHubClient, pc.Logger, pc.PluginConfig.BranchUpdater, &event)
+}
+
+// If Tide thinks this PR should be mergeable, but GitHub says that the PR's branch is "behind",
+// ask GitHub to update the branch for us.
+func handlePR(gc githubClient, log *logrus.Entry, config plugins.BranchUpdater, event *github.PullRequestEvent) error {
+	// Ignore any event subtype that we don't care about
+	if !isEventRelevant(event.Action, enabledEvents) {
+		return nil
+	}
+
+	org := event.Repo.Owner.Login
+	repo := event.Repo.Name
+	pr := event.PullRequest
+
+	// Find the Tide status.
+	tideStatus, err := findTideStatus(gc, org, repo, pr)
+	if err != nil {
+		log.WithError(err).Errorf("Failed to get the context statuses on %s/%s#%d.", org, repo, pr.Number)
+		return err
+	}
+	// Do nothing if Tide isn't in use here.
+	if tideStatus == "" {
+		log.Debugf("Skipping PR %d in repo %s/%s not monitored by Tide.", pr.Number, org, repo)
+		return nil
+	}
+
+	// And only take action if that status is success.
+	if tideStatus != tideContextStatusSuccess {
+		return nil
+	}
+
+	// If we have a "Tide mergeable" PR, check if GitHub agrees, and refresh our local PR state if we don't have an answer
+	// https://docs.github.com/en/rest/guides/getting-started-with-the-git-database-api#checking-mergeability-of-pull-requests
+	if pr.Mergable == nil {
+		// Crude, but this should always be enough time for GitHub to reach a decision.
+		// Worst case, we'll re-check the next time the PR is updated.
+		time.Sleep(gitHubMergeStateRefreshDelay * time.Second)
```

Comment:
It doesn't appear that `handlePR()` is called in an async way. Does sleeping for 10s prevent the handling of other PR events coming into prow or its plugins?

Reply:
`handlePR()` is indeed handled async. 

Plugins are embedded in the `hook` binary, and [`Server.handlePullRequestEvent()` here](https://code.i8e.io/github.com/improbable/k8s-test-infra@improbable/-/blob/prow/hook/events.go#L192) is how `hook` fires off each of the plugins that have registered for a given event type. As you can see, it fires off each one as a separate goroutine (eg `go func()`).

Doing a static 10 second sleep in this uncommon case is not going to be an issue. That said, I think it's probably a good idea to log when this does happen for debuggability.

Reply:
Log line added here 👍